File: 양진기/week10[트리]/4803.java
import java.util.*;
import java.io.*;

class Main{
    static ArrayList<ArrayList<Integer>> graph;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        int testcase = 1;

        while(true){
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());   // 정점
            int m = Integer.parseInt(st.nextToken());   // 간선

            if(n == 0 && m == 0) break;

            // 그래프 초기화
            graph = new ArrayList<>();
            for(int i=0; i<=n; i++){
                graph.add(new ArrayList<>());
            }

            // 그래프 구성
            for(int i=0; i<m; i++){
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());

                // 양방향 추가
                graph.get(a).add(b);
                graph.get(b).add(a);
            }

            visited = new boolean[n + 1];

            int count = 0;  // 트리 개수
            // 여러 트리가 존재할 수 있으므로 모든 노드를 검사해봄
            for(int i=1; i<=n; i++){
                if(!visited[i]){
                    count += checkTree(i);
                }
            }

            if(count == 0){
                sb.append("Case " + testcase++ + ": No trees.");
            } else if (count == 1){
                sb.append("Case " + testcase++ + ": There is one tree.");
            } else {
                sb.append("Case " + testcase++ + ": A forest of " + count + " trees.");
            }
            sb.append("\n");
        } // end of while(true)
        bw.write(sb.toString());
        bw.close();
        br.close();
    } // end of main

    /**
     트리의 경우 노드 개수 = 간선 개수 + 1을 활용
     양방향 간선으로 그래프를 구성하였으므로 간선 개수가 2배
     node = (edge / 2) + 1이면 트리이다.

     @param start 시작점
     @return 트리이면 1, 아니면 0
     */
    static int checkTree(int start){
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        int node = 0;
        int edge = 0;

        while(!queue.isEmpty()){
            int cur = queue.poll();
            node += 1;  // 큐에 정점이 담겼으므로 +1
            visited[cur] = true;

            for(int next : graph.get(cur)){
                edge += 1;  // 다음 정점이 있다는 것은 간선이 있다는 것 +1
                if(!visited[next]){
                    queue.offer(next);
                }
            }
        }
        return (edge / 2) + 1 == node ? 1 : 0;  // 트리면 1, 아니면 0
    }
}

File: 전유림/10주차[트리]/4803.py
import sys
from collections import deque
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

def findCycle(start):   # 사이클이 존재하는지 확인
    isCycle = False
    q = deque()
    q.append(start)    # 시작 노드를 큐에 넣고
    
    while q:
        now = q.popleft() 
        if visited[now]:    # 방문한 노드가 다시 나오면 사이클이 존재
            isCycle = True
        
        visited[now] = 1
        
        for next in tree[now]:  # 현재 노드와 연결된 노드들을 탐색
            if visited[next] == 0:  # 방문하지 않은 노드
                q.append(next)  # 큐에 넣고
                
    return isCycle


case = 0
while True:
    n, m = map(int, input().split())
    case += 1
    if n == 0 and m == 0:
        break
    tree = [[] for _ in range(n+1)]
    visited = [0]*(n+1) # 방문 여부
    cnt = 0
    
    # 양방향 매핑
    for _ in range(m):
        a, b = map(int, input().split())
        tree[a].append(b)
        tree[b].append(a)
    
    for node in range(1, n+1):
        if visited[node] == 0:  # 방문하지 않은 노드만 탐색
            if not findCycle(node): # 사이클이 없으면
                cnt += 1
    
    if cnt == 0:
        print(f'Case {case}: No trees.')
    elif cnt == 1:
        print(f'Case {case}: There is one tree.')
    else:
        print(f'Case {case}: A forest of {cnt} trees.')